core/src/main/java/org/mule/processor/chain/ProcessorIterator.java

```diff
@@ -0,0 +1,140 @@
+import org.mule.execution.MessageProcessorExecutionTemplate;
+import org.mule.processor.AbstractInterceptingMessageProcessorBase;
+import org.mule.routing.MessageFilter;
+
+import java.util.List;
+
+/**
+ * Iterates over a list of {@link org.mule.api.processor.MessageProcessor}'s processing them one by one using the result
+ * of the first processor to invoke the second and so on.  MessageProcessor implementations aside from simply iterating
+ * over processors implement rules regarding if and when iteration should stop early or even stop temporarily and be
+ * continued later.
+ * <p/>
+ * This implementation processes each {@link org.mule.api.processor.MessageProcessor} in sucession in the same thread
+ * until or processors have been invoked or one of the following is returned by a processor:
+ * <li>{@link org.mule.VoidMuleEvent}</li>
+ * <li><code>null</code></li>
+ */
+public class ProcessorIterator
+{
+
+    final protected MessageProcessorExecutionTemplate messageProcessorExecutionTemplate;
+    final protected boolean copyOnVoidEvent;
+    final protected List<MessageProcessor> processors;
+
+    protected volatile int index;
+    protected MuleEvent event;
+
+    public ProcessorIterator(MuleEvent event, List<MessageProcessor> processors,
+                             MessageProcessorExecutionTemplate messageProcessorExecutionTemplate, boolean copyOnVoidEvent)
+    {
+        this.event = event;
+        this.processors = processors;
+        this.copyOnVoidEvent = copyOnVoidEvent;
+        this.messageProcessorExecutionTemplate = messageProcessorExecutionTemplate;
+    }
+
+    public MuleEvent process() throws MessagingException
+    {
+        while (continueProcessing())
+        {
+            event = processNext();
+        }
+        return event;
+    }
+
+    protected boolean continueProcessing()
+    {
+        return index < processors.size() && event != null && !VoidMuleEvent.isVoid(event);
+    }
+
+    protected MuleEvent processNext() throws MessagingException
+    {
+        MessageProcessor processor = nextProcessor();
+
+        preProcess(processor);
+
+        if (copyOnVoidEvent && processorMayReturnVoidEvent(processor))
+        {
+            MuleEvent copy = OptimizedRequestContext.criticalSetEvent(event);
+            MuleEvent result = messageProcessorExecutionTemplate.execute(processor, event);
+            if (VoidMuleEvent.getInstance().equals(result))
+            {
+                result = copy;
+            }
+            return result;
+        }
+        else
+        {
+            return messageProcessorExecutionTemplate.execute(processor, event);
+        }
+    }
+
+    protected void preProcess(MessageProcessor processor)
+    {
+        // template method.
+    }
+
+    protected MessageProcessor nextProcessor()
+    {
+        return processors.get(index++);
+    }
+
+    private boolean processorMayReturnVoidEvent(MessageProcessor processor)
```

Comment:
This is existing method name, this method was just moved from DefaultMuleProcessorChain, until you think this is major i'd prefer to leave it as it was.